### PR TITLE
Convert gallery from mvp to mvvm

### PR DESCRIPTION
**Pull request by hieuwu**
## :rocket: Summary
Convert gallery from mvp to mvvm
Details of the analysis
https://glamorous-lung-568.notion.site/MVP-to-MVVM-Gallery-screen-analysis-fdb7ac52dfaf4f3a92d4fe0acbd1ab66?pvs=4

## :framed_picture: Screenshots:
Provide screenshots to make it visible to reviewer if possible (Optional)

| Before | After |
N/A

### CODE REVIEW

**Comment by kirillt** [comment]
@hieuwu thanks, could you also tell a bit about motivation behind these changes ⏤ what goal you want to achieve by the refactoring? Maybe pros & cons of MVP and MVVM

**Comment by hieuwu** [comment]
> @hieuwu thanks, could you also tell a bit about motivation behind these changes ⏤ what goal you want to achieve by the refactoring? Maybe pros & cons of MVP and MVVM
> 
> Am I right, that some parts of current codebase will be removed when you complete this PR? I see only additions right now

Yeah I am writing a document that includes current code execution, pros/cons & how to analyze current code then convert them to MVVM step by step. One thing I can reveal for now is that it would help us a lot in adoption to Jetpack Compose if we want in the future.

Yes, current code could be removed after we complete testing for new one.

**Comment by mdrlzy** [comment]
Nice work! We need to get rid of outdated MVP.
What do you think about using [mvi-orbit](https://github.com/orbit-mvi/orbit-mvi)?
For example you can check [FoldersViewModel](https://github.com/ARK-Builders/ARK-Navigator/blob/main/app/src/main/java/dev/arkbuilders/navigator/presentation/screen/folders/FoldersViewModel.kt)

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+    private val _deleteResource: MutableStateFlow<Int?> = MutableStateFlow(null)
+    val deleteResource: StateFlow<Int?> = _deleteResource
+
+    private val _toastIndexFailedPath: MutableStateFlow<Path?> =
+        MutableStateFlow(null)
+    val toastIndexFailedPath: StateFlow<Path?> = _toastIndexFailedPath
+
+    private val _showInfoAlert: MutableStateFlow<ShowInfoData?> =
+        MutableStateFlow(null)
+    val showInfoAlert: StateFlow<ShowInfoData?> = _showInfoAlert
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftViewModel.kt:126]: I think this is hard to read and edit, let's redo it with [mvi-orbit](https://github.com/orbit-mvi/orbit-mvi)
We need to declare a state with fields like `controlsVisible`, `currentItem`, `selectingEnabled`. And side effects such as `NavigateBack`, `ShowInfoAlert`

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/domain/ProgressWithText.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package dev.arkbuilders.navigator.presentation.screen.gallery.galleryuplift.domain
+
+data class ProgressWithText(
+    val isVisible: Boolean,
+    val text: String,
+)
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/domain/ProgressWithText.kt:6]: I made a mistake. This object is created in the viewmodel, where we cannot use string resources, but use plain text. Perhaps it’s worth making an enum with different progress states and turning it into text in fragment.

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+                        path?.let {
+                            shareResource(path)
+                        }
+                    }
+                }
+                launch {
+                    viewModel.toggleSelect.collect {
+                        toggleSelecting(it)
+                    }
+                }
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt:222]: Hard to read too

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+class GalleryUpliftFragment : Fragment() {
+    private val binding by viewBinding(FragmentGalleryBinding::bind)
+
+    @Inject
+    lateinit var factory: GalleryUpliftViewModelFactory.Factory
+    private val viewModel: GalleryUpliftViewModel by viewModels {
+        factory.create(
+            selectorNotEdit = false
+        )
+    }
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt:76]: maybe `requireArguments().getBoolean(SELECTING_ENABLED_KEY)` ?

**Review by mdrlzy — CHANGES_REQUESTED** [review]

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftViewModel.kt (hunk 1 above)
Comment (hieuwu) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftViewModel.kt]: Great. I just applied and the viewModel is a bit cleaner. Thank man

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt (hunk 3 above)
Comment (hieuwu) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt]: Replaced these with side effect

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt (hunk 4 above)
Comment (hieuwu) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt:76]: Updated

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+class GalleryUpliftFragment : Fragment() {
+    private val binding by viewBinding(FragmentGalleryBinding::bind)
+
+    @Inject
+    lateinit var factory: GalleryUpliftViewModelFactory.Factory
+    private val viewModel: GalleryUpliftViewModel by viewModels {
+        factory.create(
+            requireArguments().getBoolean(SELECTING_ENABLED_KEY)
+        )
+    }
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt:76]: Look at the [previous version](https://github.com/ARK-Builders/ARK-Navigator/blob/fc2ffe283cc9966fbe0f1098b06652fb98e82c7d/app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/GalleryFragment.kt#L64), there are clearly not enough arguments here

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+            Files.delete(path)
+        }
+        index.updateAll()
+        intent {
+            postSideEffect(GallerySideEffect.NotifyResourceChange)
+        }
+    }
+
+    private fun onFirstViewAttach() {
+        analytics.trackScreen()
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftViewModel.kt:189]: This is a name from the presenter, which does not mean anything. Maybe replace it with `initStorages`?

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/state/GalleryState.kt [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    data object NotifyTagsChanged : GallerySideEffect()
+    data class ShowEditTagsDialog(val data: ShowEditTagsData) : GallerySideEffect()
+    data class SetUpPreview(val data: SetupPreview) : GallerySideEffect()
+    data class DisplaySelectedFile(val data: DisplaySelected) : GallerySideEffect()
+    data object NotifyResourceChange : GallerySideEffect()
+    data class ShowProgressWithText(val text: ProgressWithText) : GallerySideEffect()
+    data object NotifyCurrentItemChange : GallerySideEffect()
+    data object UpdatePagerAdapterWithDiff : GallerySideEffect()
+    data class ToggleSelect(val isEnabled: Boolean) : GallerySideEffect()
```
Comment (mdrlzy) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/state/GalleryState.kt:46]: Now classes from the domain are useless. Can you remove them and move their fields to these classes?

**Review by mdrlzy — CHANGES_REQUESTED** [review]
Nice work, a little more to go. Don't forget to replace `ProgressWithText` with enum.

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/state/GalleryState.kt (hunk 7 above)
Comment (mdrlzy) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/state/GalleryState.kt:46]: For example `class DisplaySelectedFile(val data: DisplaySelected)` can be replaced with 
```
class DisplaySelectedFile(
    val selected: Boolean,
    val showAnim: Boolean,
    val selectedCount: Int,
    val itemCount: Int,
)
``` 
and `DisplaySelected` class can be removed

**Comment by shubertm** [comment]
https://github.com/ARK-Builders/ARK-Navigator/assets/87703131/67d43ce7-7d47-49b8-9d8c-c2986308eede

- [ ] Tag shows on wrong resource in Gallery
- [ ] Selection goes forward by 1 step or more when scrolling in Gallery

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt (hunk 5 above)
Comment (hieuwu) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/gallery/galleryuplift/GalleryUpliftFragment.kt:76]: Actually I extracted these parameters out of the constructors and put in `initlialize` method. But it looks like that's not the proper way, let me move them back to constructor

**Comment by hieuwu** [comment]
> VID_20240613_161155.mp4 
> * [x]  Tag shows on wrong resource in Gallery
> * [x]  Selection goes forward by 1 step or more when scrolling in Gallery

Thank you for pointing them out, I have fixed them. Please help check again.

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/resources/adapter/ResourcesGridPresenter.kt [hunk 8]
```diff
@@ -145,8 +145,16 @@ class ResourcesGridPresenter(
                 return@launch
         }
 
+//        router.navigateToFragmentUsingAdd(
```
Comment (shubertm) [inline app/src/main/java/dev/arkbuilders/navigator/presentation/screen/resources/adapter/ResourcesGridPresenter.kt:148]: Remove commented out code

Comment (hieuwu) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/resources/adapter/ResourcesGridPresenter.kt:148]: This package will be removed after I merge this PR. No worries

**Review by shubertm — COMMENTED** [review]
Great work :+1:

File: app/src/main/java/dev/arkbuilders/navigator/presentation/screen/resources/adapter/ResourcesGridPresenter.kt (hunk 8 above)
Comment (kirillt) [reply app/src/main/java/dev/arkbuilders/navigator/presentation/screen/resources/adapter/ResourcesGridPresenter.kt:148]: @hieuwu I think it's better to remove comments before merging. If we want to restore older version we can just checkout older commit..

**Review by mdrlzy — APPROVED** [review]

**Pull request merged**